README.md
# markdown-it-github-headings

Add GitHub style anchor tags to headers

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![standard][standard-image]][standard-url]

[npm-image]: https://img.shields.io/npm/v/markdown-it-github-headings.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/markdown-it-github-headings
[travis-image]: https://img.shields.io/travis/Flet/markdown-it-github-headings.svg?style=flat-square
[travis-url]: https://travis-ci.org/Flet/markdown-it-github-headings
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[standard-url]: http://npm.im/standard

## Install

```
npm install markdown-it-github-headings
```

## Usage

```js
var md = require('markdown-it')()
  .use(require('markdown-it-github-headings'), options)

```

## Options and Defaults

The defaults will make the heading anchors behave as close to how GitHub behaves as possible.

Name              | Description                                                    | Default
------------------|----------------------------------------------------------------|-----------------------------------
`className`                     | name of the class that will be added to the anchor tag               | `anchor`
`prefixHeadingIds`              | add a prefix to each heading ID. *(see security note below)*         | `true`
`prefix`                        | if `prefixHeadingIds` is true, use this string to prefix each ID.    | `user-content-`
`enableHeadingLinkIcons`        | Adds the icon <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg> next to each heading | `true`
`linkIcon`                      | If `enableHeadingLinkIcons` is true, use this to supply a custom icon (or anything really) | <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>
`resetSlugger`                     | reset the slugger counter between .render calls for duplicate headers. (See tests for example)              | true

## Why should I prefix heading IDs?
When using user generated content, its possible to run into **DOM Clobbering** when heading IDs are generated. Since IDs are used by JavaScript and CSS, a user could craft a page that breaks functionality or styles. A good way to avoid clobbering is to add a prefix to every generated ID to ensure they cannot overlap with existing IDs.

If you have full control over the content, there is less of a risk, but be aware that strange bugs related to DOM Clobbering are still possible!

For more information, here are some good resources on the topic:
- [User-generated content and DOM clobbering](http://opensoul.org/2014/09/05/dom-clobbering/)
- [In the DOM, no one will hear you scream](https://www.slideshare.net/x00mario/in-the-dom-no-one-will-hear-you-scream)
- [A discussion about GitHub implementation](https://github.com/jch/html-pipeline/pull/111#issuecomment-34369984)
- [An open issue on markdown-it repo](https://github.com/markdown-it/markdown-it/issues/28)

### But the prefixes make links look real gross.
One solution is to write some client side JavaScript to force non-prefixed hashes to jump to prefixed anchors. This is how its handled on GitHub and npmjs.com.

Check out [marky-deep-links](https://github.com/Flet/marky-deep-links) for an example (works great with browserify or webpack).


## Contributing

Contributions welcome! Please read the [contributing guidelines](CONTRIBUTING.md) first.

## License

[ISC](LICENSE.md)


CHANGELOG.md
# markdown-it-github-headings change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## 2.0.1 2022-06-09

 - Add TypeScript Declaration #19 Thank you [@ashaifarhan](https://github.com/ashalfarhan) for contributing!
 - Updated dev dependencies to squelch audit warnings.

## 2.0.0 2020-01-14

  - BREAKING: add resetSlugger option and default to true
    - This change avoids needlessly incrementing the duplicate slug counter between .render() calls. It is now the default behavior.
    - set the option `resetSlugger` to `false` to switch back to the old behavior if desired.

## 1.1.2 2019-09-23

  - Update dependencies per `npm audit`

## 1.1.1 2018-02-27

  - Removed dependency on `fs` for browser compatibility. Thanks [@imcuttle](https://github.com/imcuttle)!

# 1.1.0 2017-11-18
  - Added option to allow a custom svg icon for links (Thanks @adam-lynch!).

## 1.0.0 2017-04-28

Released!


test/index.js
const test = require('tape')

const md = require('markdown-it')
const anchor = require('..')
const svg = '<svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>'

test('works as expected', function (t) {
  const result = md().use(anchor).render('# Hello')
  const expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('resets slugger on every run', function (t) {
  const instance = md().use(anchor)

  let result = instance.render('# Hello')
  let expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')

  result = instance.render('# Hello')
  expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')

  t.end()
})

test('resetSlugger false does not reset on every run', function (t) {
  const instance = md().use(anchor, { resetSlugger: false })

  const result = instance.render('# Hello')
  const expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')

  const result2 = instance.render('# Hello')
  const expectedResult2 = `<h1><a id="user-content-hello-1" class="anchor" href="#hello-1" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result2, expectedResult2, 'works')

  t.end()
})

test('multiple headers', function (t) {
  const input = `
## Hello
### World
`

  const expectedResult = `<h2><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h2>
<h3><a id="user-content-world" class="anchor" href="#world" aria-hidden="true">${svg}</a>World</h3>
`
  const result = md().use(anchor).render(input)
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('duplicate headers', function (t) {
  const input = `
# Hello
## Hello
### Hello
`

  const expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>
<h2><a id="user-content-hello-1" class="anchor" href="#hello-1" aria-hidden="true">${svg}</a>Hello</h2>
<h3><a id="user-content-hello-2" class="anchor" href="#hello-2" aria-hidden="true">${svg}</a>Hello</h3>
`
  const result = md().use(anchor).render(input)
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('prefix: foo-', function (t) {
  const result = md().use(anchor, { prefix: 'foo-' }).render('# Hello')
  const expectedResult = `<h1><a id="foo-hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('prefixHeadingIds: false', function (t) {
  const result = md().use(anchor, { prefixHeadingIds: false }).render('# Hello')
  const expectedResult = `<h1><a id="hello" class="anchor" href="#hello" aria-hidden="true">${svg}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('enableHeadingLinkIcons: false', function (t) {
  const result = md().use(anchor, { enableHeadingLinkIcons: false }).render('# Hello')
  const expectedResult = '<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true"></a>Hello</h1>\n'
  t.equals(result, expectedResult, 'works')
  t.end()
})

test('custom linkIcon', function (t) {
  const customContents = '<svg xmlns="http://www.w3.org/2000/svg" width="15" height="15" viewBox="0 0 8 8"><path d="M5.88.03c-.18.01-.36.03-.53.09-.27.1-.53.25-.75.47a.5.5 0 1 0 .69.69c.11-.11.24-.17.38-.22.35-.12.78-.07 1.06.22.39.39.39 1.04 0 1.44l-1.5 1.5c-.44.44-.8.48-1.06.47-.26-.01-.41-.13-.41-.13a.5.5 0 1 0-.5.88s.34.22.84.25c.5.03 1.2-.16 1.81-.78l1.5-1.5c.78-.78.78-2.04 0-2.81-.28-.28-.61-.45-.97-.53-.18-.04-.38-.04-.56-.03zm-2 2.31c-.5-.02-1.19.15-1.78.75l-1.5 1.5c-.78.78-.78 2.04 0 2.81.56.56 1.36.72 2.06.47.27-.1.53-.25.75-.47a.5.5 0 1 0-.69-.69c-.11.11-.24.17-.38.22-.35.12-.78.07-1.06-.22-.39-.39-.39-1.04 0-1.44l1.5-1.5c.4-.4.75-.45 1.03-.44.28.01.47.09.47.09a.5.5 0 1 0 .44-.88s-.34-.2-.84-.22z"></path></svg><img/><h1>Hello<span><span>World</span></span></h1>'
  const result = md().use(anchor, { linkIcon: customContents }).render('# Hello')
  const expectedResult = `<h1><a id="user-content-hello" class="anchor" href="#hello" aria-hidden="true">${customContents}</a>Hello</h1>\n`
  t.equals(result, expectedResult, 'works')
  t.end()
})


index.d.ts
declare module 'markdown-it-github-headings' {
  import type { PluginWithOptions } from 'markdown-it';
  type GithubHeadingsPluginOptions = {
    /**
     * Name of the class that will be added to the anchor tag.
     * @default "anchor"
     */
    className?: string;
    /**
     * Add a prefix to each heading ID.
     * @see https://github.com/Flet/markdown-it-github-headings#why-should-i-prefix-heading-ids
     * @default true
     */
    prefixHeadingIds?: boolean;
    /**
     * If `prefixHeadingIds` is true, use this string to prefix each ID.
     * @default "user-content"
     */
    prefix?: string;
    /**
     * Adds the icon next to each heading.
     * @default true
     */
    enableHeadingLinkIcons?: boolean;
    /**
     * If `enableHeadingLinkIcons` is true, use this to supply a custom icon (or anything really).
     */
    linkIcon?: string;
    /**
     * Reset the slugger counter between .render calls for duplicate headers. (See tests for example).
     * @default true
     */
    resetSlugger?: boolean;
  };
  const markdownItGithubHeadings: PluginWithOptions<GithubHeadingsPluginOptions>;
  export = markdownItGithubHeadings;
}


index.js
module.exports = plugin

const GithubSlugger = require('github-slugger')
const innertext = require('innertext')
const defaultOptions = {
  enableHeadingLinkIcons: true,
  prefixHeadingIds: true,
  prefix: 'user-content-',
  className: 'anchor',
  // shamelessly borrowed from GitHub, thanks y'all
  linkIcon: '<svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>',
  resetSlugger: true,
  headingLinksHref: slug => `#${slug}`
}

function plugin (md, _opts) {
  const options = Object.assign({}, defaultOptions, _opts)

  if (!options.prefixHeadingIds) options.prefix = ''

  const slugger = new GithubSlugger()
  let Token

  md.core.ruler.push('headingLinks', function (state) {
    if (options.resetSlugger) {
      slugger.reset()
    }

    // save the Token constructor because we'll be building a few instances at render
    // time; that's sort of outside the intended markdown-it parsing sequence, but
    // since we have tight control over what we're creating (a link), we're safe
    if (!Token) {
      Token = state.Token
    }
  })

  md.renderer.rules.heading_open = function (tokens, idx, opts, env, self) {
    const children = tokens[idx + 1].children
    // make sure heading is not empty
    if (children && children.length) {
      // Generate an ID based on the heading's innerHTML; first, render without
      // converting gemoji strings to unicode emoji characters
      const unemojiWithToken = unemoji.bind(null, Token)
      const rendered = md.renderer.renderInline(children.map(unemojiWithToken), opts, env)
      const postfix = slugger.slug(
        innertext(rendered)
          .replace(/[<>]/g, '') // In case the heading contains `<stuff>`
          .toLowerCase() // because `slug` doesn't lowercase
      )

      // add 3 new token objects link_open, text, link_close
      const linkOpen = new Token('link_open', 'a', 1)
      const text = new Token('html_inline', '', 0)
      if (options.enableHeadingLinkIcons) {
        text.content = options.linkIcon
      }
      const linkClose = new Token('link_close', 'a', -1)

      // add some link attributes
      linkOpen.attrSet('id', options.prefix + postfix)
      linkOpen.attrSet('class', options.className)
      linkOpen.attrSet('href', options.headingLinksHref(postfix))
      linkOpen.attrSet('aria-hidden', 'true')

      // add new token objects as children of heading
      children.unshift(linkClose)
      children.unshift(text)
      children.unshift(linkOpen)
    }

    return md.renderer.renderToken(tokens, idx, options, env, self)
  }
}

function unemoji (TokenConstructor, token) {
  if (token.type === 'emoji') {
    return Object.assign(new TokenConstructor(), token, { content: token.markup })
  }
  return token
}
